server/src/main/java/org/opensearch/common/regex/Regex.java

```diff
@@ -93,6 +93,23 @@ public static Automaton simpleMatchToAutomaton(String... patterns) {
         return Operations.union(automata);
     }
 
+    /**
+     *
+     * @param str - The input string to remove adjacent duplicate characters from
+     * @param target - The target character to remove duplicates of
+     * @return - string with adjacent duplicates of the target character removed
+     */
+    static String removeAdjacentDuplicates(String str, char target) {
```

Comment:
Won't the simple change in `simpleMatchWithNormalizedStrings` fix the problem by skipping over the `*` sequence? 

```
            } else if (nextIndex == 1) {
                // The wildcard series, "**" - skipping all till the last "*"
                int skipTillIndex = nextIndex + 1;
                while (pattern.length() < skipTillIndex && pattern.charAt(skipTillIndex) == '*') { 
                    ++skipTillIndex;
                };
                return simpleMatchWithNormalizedStrings(pattern.substring(skipTillIndex - 1), str);
            }
```

Reply:
Either approach is fine. I chose this approach aiming to ensure that the input to the `simpleMatchWithNormalizedStrings` was valid so that the logic for this edge case did not need to reside within the method. It could be considered part of the `normalization` of the input.

Reply:
The `simpleMatchWithNormalizedStrings` for some reasons accepted only `**` but not the sequences of it, modify the regex in this cases basically is possible but there is clearly more than one place now that deals with `*`, consolidating the logic in a single place seems to make sense.

Reply:
This PR performs the [normalization](https://github.com/opensearch-project/OpenSearch/blob/57ff3cdf6ae186218983df6bd3a79e4a3f8bcd55/server/src/main/java/org/opensearch/common/regex/Regex.java#L138-L145) before feeding the normalized input into `simpleMatchWithNormalizedStrings`. There is currently only one place where `simpleMatchWithNormalizedStrings` is called. I chose to put the logic before the invocation of `simpleMatchWithNormalizedStrings` so that its only called once before and not on each recurse of the call. 

Do you think it makes sense to put the call elsewhere to normalize the input?

Reply:
> Do you think it makes sense to put the call elsewhere to normalize the input?

I see 2 issues with normalization approach:
- if you look for the exception that `simpleMatchWithNormalizedStrings` throws now, it prints out the **normalized** pattern and not what user has really sent out (`trimmedPattern` vs `pattern`), we are losing the truth source here
- we do double work since we still have to go and parse whole pattern (but we do this twice now, one time to fold the `*` and second time to actually go and parse it again)

It seemed to me that folding `*` in one go is simpler and more straightforward approach, and it does not require pre-normalization call.

Reply:
> if you look for the exception that simpleMatchWithNormalizedStrings throws now, it prints out the normalized pattern and not what user has really sent out (trimmedPattern vs pattern), we are losing the truth source here

This seems solvable by passing original pattern

>we do double work since we still have to go and parse whole pattern (but we do this twice now, one time to fold the * and second time to actually go and parse it again)

This would add a complexity of m making it O(n+m) with n>=m, where in worst case scenario where there are no `*`s, m= n leading to O(2n) which would still be constant time.

> It seemed to me that folding * in one go is simpler and more straightforward approach, and it does not require pre-normalization call.

This could be followed if its guaranteed that `simpleMatchWithNormalizedStrings` will be called only once

Reply:
The mere presence of the "simple match" optimization indicates it's intended for performance-sensitive/hot-path applications to shortcut regex.

I had originally started a comment saying that this approach was going to create another String just to temporarily toss it onto the garbage heap, but then I realized the entire method is recursive for every wildcard (even non-adjacent ones).    So the String "opensearch*is*an*awesome*search*engine" will recurse 5 times, generating more trash to be GC'd.  If you're taking the effort to eliminate one recursion in "**" let's go all the way and eliminate all the recursion.

Your one-pass method here is a step in the right direction but you can do even better by doing a single pass over each of the two strings without ever creating any new String objects.

Suggestion:
 - Keep the original `pattern` and `str` objects, never create new ones (or substrings)
 - Keep an `int` starting index pointer for each of the two strings
 - Use `indexOf()` to find the wildcard position and `regionMatches()` to compare up until that point 
 - Use `indexOf()` overload to find the next index (replace the if/then checking only one dupe with a loop until wildcards are done as @reta suggested)
 - Use `substring()` to find the bits in between these indices (current temporary `part` String, don't think there's an easy way around that) and the position of that in str
 - repeat

That suggestion mostly follows the existing code except for adding @reta's improvement and replacing the final recursive line with a loop construct where you use the index pointers instead of 0.

 - 
1. Don't use `str.toCharArray()` in your iteration, that's another creation of a new (array) object just to throw it away.  Keep the original strings and just iterate using an integer and `charAt()`.
2. With a com

Reply:
OK, so I have just spent most of the last hour trying to implement the approach I outlined above.  And it just ain't worth the sacrifice of readability for the slight improvement in efficiency.

I did, in my various iterations, come upon what I think is an elegant solution which is much more readable than the current recursive mess:

1. Alter this method to return a `List<String>` instead of just a String.  Do this by:
   - iterate by character and adding to a StringBuilder (except use `charAt()` other than an array)
   - when you encounter a wildcard, add the existing string to the list if it's not empty, and reset the builder
2. In the `simpleMatch()` method just track a "position" in the string, starting at 0
3. Iterate your list and find str.indexOf(substringfromlist, startpos) advancing the str pointer each time.  Pos should start at exactly 0 (unless special case of pattern starting with *) otherwise >0.
4. Advance to that index + length.   Repeat iteration from 3.
5. Special case the ends-with wildcard.

Reply:
Thank you @dbwiddis. I updated the algorithm to remove recursion entirely. It now uses an iterative 2 pointer-style solution.